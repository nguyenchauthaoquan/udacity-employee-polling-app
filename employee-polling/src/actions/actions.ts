import {
    AUTHENTICATE_USER,
    GET_QUESTIONS,
    GET_USERS,
    LOGOUT,
    SAVE_ANSWERS_QUESTION,
    SAVE_QUESTION, SAVE_USER_ANSWERS, SAVE_USER_QUESTION
} from "../constants/constants.ts";
import {Action} from "../models/components/action.ts";
import {Dispatch} from "redux";
import {_getQuestions, _getUsers, _saveQuestion, _saveQuestionAnswer} from "../utils/_DATA.ts";
import User from "../models/data/user.ts";
import {reducer} from "../reducers/reducer.ts";
import Question from "../models/data/question.ts";

export const authenticateUserAction = (payload: object): Action => {
    return {
        type: AUTHENTICATE_USER,
        payload
    }
}

export const logoutAction = (): Action => {
    return {
        type: LOGOUT,
    }
}

export const getUsersAction = (payload: object): Action => {
    return {
        type: GET_USERS,
        payload
    }
}

export const getQuestions = (payload: object) => {
    return {
        type: GET_QUESTIONS,
        payload
    }
}

export const saveUserQuestionAction = (payload: object) => {
    return {
        type: SAVE_USER_QUESTION,
        payload
    }
}

export const saveUserAnswerAction = (payload: object) => {
    console.log(payload)
    return {
        type: SAVE_USER_ANSWERS,
        payload,
    }
}

export const saveQuestionAction = (payload: Question) => {
    return {
        type: SAVE_QUESTION,
        payload,
    }
}

export const saveAnswerQuestionAction = (payload: object) => {
    return {
        type: SAVE_ANSWERS_QUESTION,
        payload,
    }
}

export const getAppData = (): (dispatch: Dispatch<Action>) => Promise<void> => {
    return async (dispatch: Dispatch<Action>) => {
        const [users, questions] = await Promise.all([_getUsers(), _getQuestions()]);
        const {users: users_1, questions: questions_1} = ({
            users,
            questions
        });
        dispatch(getQuestions(questions_1));
        dispatch(getUsersAction(users_1));
    }
}

export function handleLogin(username: string, password: string) {
    return (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const { users} = getState();

        const user = Object.values(users as User).find((user: User) => user.id === username && user.password === password);

        if (user) {
            return dispatch(authenticateUserAction(user));
        }
    };
}

export function handleLogout(): (dispatch: Dispatch<Action>) => Action {
    return (dispatch: Dispatch<Action>) => {
        return dispatch(logoutAction());
    };
}

export function handleSaveQuestion(optionOne: string, optionTwo: string) {
    return async (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const {authenticatedUser} = getState();
        const user = authenticatedUser as User
        const question = await _saveQuestion({
            optionOneText: optionOne,
            optionTwoText: optionTwo,
            author: user
        }) as Question;

        dispatch(saveQuestionAction(question));
        dispatch(saveUserQuestionAction({
            author: question.author,
            id: question.id
        }));

    }
}

export function handleSaveAnswerQuestion(questionId: string, answer: string) {
    return (dispatch: Dispatch<Action>, getState: () => ReturnType<typeof reducer>) => {
        const {authenticatedUser} = getState();
        const user: User = authenticatedUser as User;

        _saveQuestionAnswer({
            authedUser: user.id,
            qid: questionId,
            answer: answer
        }).then(() => {
            dispatch(saveAnswerQuestionAction({
                author: user.id,
                qid: questionId,
                answer: answer
            }));
            dispatch(saveUserAnswerAction({
                authedUser: user.id,
                qid: questionId,
                answer: answer
            }))
        })
    }
}