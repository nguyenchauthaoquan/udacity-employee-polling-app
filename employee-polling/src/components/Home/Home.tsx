import {FC} from 'react';
import {DashBoardProps, HomeProps} from "../../models/components/props.ts";
import DashBoard from "../DashBoard/DashBoard.tsx";
import {connect} from "react-redux";
import {reducer} from "../../reducers/reducer.ts";
import User from "../../models/data/user.ts";
import Question from "../../models/data/question.ts";

const Home: FC<HomeProps> = () => {
    return (
        <DashBoard  />
    );
}

const mapStateToProps = (state: ReturnType<typeof reducer>): DashBoardProps => {
    return {
        user: state.authenticatedUser as User,
        users: state.users,
        questions: Object.values(state.questions as Question).sort((a, b) => b.timestamp - a.timestamp)
    }
}

export default connect(mapStateToProps)(Home);