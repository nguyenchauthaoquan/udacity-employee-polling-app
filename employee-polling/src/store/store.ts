import {configureStore} from '@reduxjs/toolkit';
import {reducer} from "../reducers/reducer.ts";

export const store = configureStore(
    {
        reducer: reducer
    }
)