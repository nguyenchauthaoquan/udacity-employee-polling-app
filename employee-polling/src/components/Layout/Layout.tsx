import {FC} from 'react';
import {LayoutProps} from "../../models/components/props.ts";
import {Button, Col, Container, Nav, Row} from "react-bootstrap";
import {reducer} from "../../reducers/reducer.ts";
import User from "../../models/data/user.ts";
import {ThunkDispatch} from "redux-thunk";
import {Action} from "../../models/components/action.ts";
import {handleLogout} from "../../actions/actions.ts";
import {connect} from "react-redux";
import {Link, useNavigate} from "react-router-dom";

const Layout: FC<LayoutProps> = (props: LayoutProps) => {
    const navigate = useNavigate();

    const handleLogout = () => {
        props.logout && props.logout();

        navigate("/")

    }

    return (
        <Container>
            {props.user && (
                <>
                    <Row>
                        <Col md={6}>
                            <Nav variant={"underline"} activeKey={window.location.pathname} data-testid="navbar">
                                <Nav.Item>
                                    <Link to={"/home"}>Home</Link>
                                </Nav.Item>
                                <Nav.Item>
                                    <Link to={"/leaderboard"}>Leaderboard</Link>
                                </Nav.Item>
                                <Nav.Item>
                                    <Link to={"/new"}>New</Link>
                                </Nav.Item>
                            </Nav>
                        </Col>
                        <Col md={6} className={"d-flex justify-content-end"}>
                            <Nav>
                                <Nav.Item>
                                    <Nav.Link disabled data-testid="username">
                                        {props.user?.id}
                                    </Nav.Link>
                                </Nav.Item>
                                <Nav.Item>
                                    <Button data-testid="logout-btn" variant={"secondary"} onClick={() => handleLogout()}>Log out</Button>
                                </Nav.Item>
                            </Nav>
                        </Col>
                    </Row>
                    <Row>
                        <Col md={12}>
                            <hr />
                        </Col>
                    </Row>
                </>
            )}
            <Row>
                <Col md={12}>
                    {props.children}
                </Col>
            </Row>
        </Container>
    );
};


const mapStateToProps = (state: ReturnType<typeof reducer>, props: LayoutProps)=> {
    return {
        ...props,
        user: state.authenticatedUser as User
    }
}

const mapDispatchToProps = (dispatch: ThunkDispatch<ReturnType<typeof reducer>, never, Action>) => {
    return {
        logout: () => {
            dispatch(handleLogout());
        }
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(Layout);