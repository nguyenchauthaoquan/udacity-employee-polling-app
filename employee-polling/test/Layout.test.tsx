import {fireEvent, render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../src/store/store";
import {BrowserRouter} from "react-router-dom";
import Layout from "../src/components/Layout/Layout";
import {authenticateUserAction, getAppData} from "../src/actions/actions";
import Home from "../src/components/Home/Home";
import LeaderBoard from "../src/components/LeaderBoard/LeaderBoard";
import NewPoll from "../src/components/NewPoll/NewPoll";

describe("Layout Component", () => {
    it("should be rendered successfully", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    })
    it ("should have the navbar", () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout/>
                </BrowserRouter>
            </Provider>
        );

        const navbarElement = component.getByTestId("navbar");

        expect(component).toBeDefined();
        expect(navbarElement).toBeInTheDocument();
    })
    it("should display the username", () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout/>
                </BrowserRouter>
            </Provider>
        );

        const usernameElement = component.getByTestId("username");

        expect(component).toBeDefined();
        expect(usernameElement).toBeInTheDocument();
        expect(usernameElement.textContent).toBe("sarahedo")
    })
    it ("should have the logout button", async () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout/>
                </BrowserRouter>
            </Provider>
        );

        const logoutElement = component.getByTestId("logout-btn");
        expect(component).toBeDefined();
        expect(logoutElement).toBeInTheDocument();
        expect(logoutElement.textContent).toBe("Log out")
    })

    it ("should click the logout button successfully", async () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout/>
                </BrowserRouter>
            </Provider>
        );

        const logoutElement = component.getByTestId("logout-btn");
        expect(component).toBeDefined();
        expect(logoutElement).toBeInTheDocument();
        expect(logoutElement.textContent).toBe("Log out")
        fireEvent.click(logoutElement)
    })

})

describe("Home component", () => {
    it("should render the dashboard component", () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        store.dispatch(getAppData())

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout>
                        <Home />
                    </Layout>
                </BrowserRouter>
            </Provider>
        );
        const dashboardElement = component.getByTestId("dashboard")

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
        expect(dashboardElement).toBeInTheDocument();
    })
})

describe("Leaderboard Component", async () => {
    it("should render the leaderboard table", () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        store.dispatch(getAppData())

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout>
                        <LeaderBoard />
                    </Layout>
                </BrowserRouter>
            </Provider>
        );
        const leaderboardElement = component.getByTestId("leaderboard-table")

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
        expect(leaderboardElement).toBeInTheDocument();
    })
})

describe("NewPoll Component", () => {
    it("should render the create poll form", async () => {
        store.dispatch(authenticateUserAction({id: "sarahedo", password: "123456"}))
        store.dispatch(getAppData())

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Layout>
                        <NewPoll />
                    </Layout>
                </BrowserRouter>
            </Provider>
        );

        const newPollElement = component.getByTestId("new-poll")

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
        expect(newPollElement).toBeInTheDocument();

        const firstOptionLabelElement = component.getByTestId("first-option-label");
        const firstOptionInputElement = component.getByTestId("first-option-input");
        const secondOptionLabelElement = component.getByTestId("second-option-label");
        const secondOptionInputElement = component.getByTestId("second-option-input");
        const submitButtonElement = component.getByTestId("poll-submit-btn");

        expect(firstOptionLabelElement).toBeInTheDocument();
        expect(firstOptionLabelElement.textContent).toBe("First Option");
        fireEvent.change(firstOptionInputElement, {target: {value: 'A'}});
        expect(firstOptionInputElement.value).toBe("A");

        expect(secondOptionLabelElement).toBeInTheDocument();
        expect(secondOptionLabelElement.textContent).toBe("Second Option");
        fireEvent.change(secondOptionInputElement, {target: {value: 'B'}});
        expect(secondOptionInputElement.value).toBe("B");

        expect(submitButtonElement).toBeInTheDocument();
        expect(submitButtonElement.textContent).toBe("Submit");
        fireEvent.click(submitButtonElement)
    })
})