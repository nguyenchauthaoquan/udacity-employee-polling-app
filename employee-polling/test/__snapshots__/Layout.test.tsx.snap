// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home component > should render the dashboard component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <div
              class="nav nav-underline"
              data-testid="navbar"
            >
              <div
                class="nav-item"
              >
                <a
                  href="/home"
                >
                  Home
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/add"
                >
                  New
                </a>
              </div>
            </div>
          </div>
          <div
            class="d-flex justify-content-end col-md-6"
          >
            <div
              class="nav"
            >
              <div
                class="nav-item"
              >
                <a
                  aria-disabled="true"
                  class="nav-link disabled"
                  data-testid="username"
                  role="button"
                >
                  sarahedo
                </a>
              </div>
              <div
                class="nav-item"
              >
                <button
                  class="btn btn-secondary"
                  data-testid="logout-btn"
                  type="button"
                >
                  Log out
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <hr />
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <div
              class="row"
              data-testid="dashboard"
            >
              <div
                class="col-md-12"
              >
                <div
                  class="card"
                >
                  <div
                    class="text-center card-header"
                  >
                    <div
                      class="card-title h5"
                    >
                      Unanswered Questions
                    </div>
                  </div>
                  <div
                    class="card-body"
                  >
                    <div
                      class="container"
                    >
                      <div
                        class="row"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="mt-3 row"
            >
              <div
                class="col-md-12"
              >
                <div
                  class="card"
                >
                  <div
                    class="text-center card-header"
                  >
                    <div
                      class="card-title h5"
                    >
                      Answered Questions
                    </div>
                  </div>
                  <div
                    class="card-body"
                  >
                    <div
                      class="container"
                    >
                      <div
                        class="row"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="nav nav-underline"
            data-testid="navbar"
          >
            <div
              class="nav-item"
            >
              <a
                href="/home"
              >
                Home
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/add"
              >
                New
              </a>
            </div>
          </div>
        </div>
        <div
          class="d-flex justify-content-end col-md-6"
        >
          <div
            class="nav"
          >
            <div
              class="nav-item"
            >
              <a
                aria-disabled="true"
                class="nav-link disabled"
                data-testid="username"
                role="button"
              >
                sarahedo
              </a>
            </div>
            <div
              class="nav-item"
            >
              <button
                class="btn btn-secondary"
                data-testid="logout-btn"
                type="button"
              >
                Log out
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <hr />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="row"
            data-testid="dashboard"
          >
            <div
              class="col-md-12"
            >
              <div
                class="card"
              >
                <div
                  class="text-center card-header"
                >
                  <div
                    class="card-title h5"
                  >
                    Unanswered Questions
                  </div>
                </div>
                <div
                  class="card-body"
                >
                  <div
                    class="container"
                  >
                    <div
                      class="row"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mt-3 row"
          >
            <div
              class="col-md-12"
            >
              <div
                class="card"
              >
                <div
                  class="text-center card-header"
                >
                  <div
                    class="card-title h5"
                  >
                    Answered Questions
                  </div>
                </div>
                <div
                  class="card-body"
                >
                  <div
                    class="container"
                  >
                    <div
                      class="row"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Layout Component > should be rendered successfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-12"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-12"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Leaderboard Component > should render the leaderboard table 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <div
              class="nav nav-underline"
              data-testid="navbar"
            >
              <div
                class="nav-item"
              >
                <a
                  href="/home"
                >
                  Home
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/add"
                >
                  New
                </a>
              </div>
            </div>
          </div>
          <div
            class="d-flex justify-content-end col-md-6"
          >
            <div
              class="nav"
            >
              <div
                class="nav-item"
              >
                <a
                  aria-disabled="true"
                  class="nav-link disabled"
                  data-testid="username"
                  role="button"
                >
                  sarahedo
                </a>
              </div>
              <div
                class="nav-item"
              >
                <button
                  class="btn btn-secondary"
                  data-testid="logout-btn"
                  type="button"
                >
                  Log out
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <hr />
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-striped table-bordered table-hover"
                data-testid="leaderboard-table"
              >
                <thead>
                  <tr>
                    <th>
                      Users
                    </th>
                    <th>
                      Answered
                    </th>
                    <th>
                      Created
                    </th>
                  </tr>
                </thead>
                <tbody />
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="nav nav-underline"
            data-testid="navbar"
          >
            <div
              class="nav-item"
            >
              <a
                href="/home"
              >
                Home
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/add"
              >
                New
              </a>
            </div>
          </div>
        </div>
        <div
          class="d-flex justify-content-end col-md-6"
        >
          <div
            class="nav"
          >
            <div
              class="nav-item"
            >
              <a
                aria-disabled="true"
                class="nav-link disabled"
                data-testid="username"
                role="button"
              >
                sarahedo
              </a>
            </div>
            <div
              class="nav-item"
            >
              <button
                class="btn btn-secondary"
                data-testid="logout-btn"
                type="button"
              >
                Log out
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <hr />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="table-responsive"
          >
            <table
              class="table table-striped table-bordered table-hover"
              data-testid="leaderboard-table"
            >
              <thead>
                <tr>
                  <th>
                    Users
                  </th>
                  <th>
                    Answered
                  </th>
                  <th>
                    Created
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`NewPoll Component > should render the create poll form 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <div
              class="nav nav-underline"
              data-testid="navbar"
            >
              <div
                class="nav-item"
              >
                <a
                  href="/home"
                >
                  Home
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </div>
              <div
                class="nav-item"
              >
                <a
                  href="/add"
                >
                  New
                </a>
              </div>
            </div>
          </div>
          <div
            class="d-flex justify-content-end col-md-6"
          >
            <div
              class="nav"
            >
              <div
                class="nav-item"
              >
                <a
                  aria-disabled="true"
                  class="nav-link disabled"
                  data-testid="username"
                  role="button"
                >
                  sarahedo
                </a>
              </div>
              <div
                class="nav-item"
              >
                <button
                  class="btn btn-secondary"
                  data-testid="logout-btn"
                  type="button"
                >
                  Log out
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <hr />
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <div
              class="container"
              data-testid="new-poll"
            >
              <div
                class="row"
              >
                <div
                  class="text-center col-md-12"
                >
                  <h3>
                    Would you rather
                  </h3>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="text-center col-md-12"
                >
                  <span
                    class="text-secondary"
                  >
                    Create Your Own Poll
                  </span>
                </div>
              </div>
              <div
                class="row"
              >
                <div
                  class="col-md-12"
                >
                  <form
                    class=""
                  >
                    <div
                      class="mb-3"
                    >
                      <label
                        class="form-label"
                        data-testid="first-option-label"
                        for="first-option"
                      >
                        First Option
                      </label>
                      <input
                        class="form-control"
                        data-testid="first-option-input"
                        id="first-option"
                        placeholder="Option One"
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      class="mb-3"
                    >
                      <label
                        class="form-label"
                        data-testid="second-option-label"
                        for="second-option"
                      >
                        Second Option
                      </label>
                      <input
                        class="form-control"
                        data-testid="second-option-input"
                        id="second-option"
                        placeholder="Option Two"
                        type="text"
                        value=""
                      />
                    </div>
                    <button
                      class="btn btn-light"
                      data-testid="poll-submit-btn"
                      type="submit"
                    >
                      Submit
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="nav nav-underline"
            data-testid="navbar"
          >
            <div
              class="nav-item"
            >
              <a
                href="/home"
              >
                Home
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                href="/add"
              >
                New
              </a>
            </div>
          </div>
        </div>
        <div
          class="d-flex justify-content-end col-md-6"
        >
          <div
            class="nav"
          >
            <div
              class="nav-item"
            >
              <a
                aria-disabled="true"
                class="nav-link disabled"
                data-testid="username"
                role="button"
              >
                sarahedo
              </a>
            </div>
            <div
              class="nav-item"
            >
              <button
                class="btn btn-secondary"
                data-testid="logout-btn"
                type="button"
              >
                Log out
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <hr />
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div
            class="container"
            data-testid="new-poll"
          >
            <div
              class="row"
            >
              <div
                class="text-center col-md-12"
              >
                <h3>
                  Would you rather
                </h3>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="text-center col-md-12"
              >
                <span
                  class="text-secondary"
                >
                  Create Your Own Poll
                </span>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col-md-12"
              >
                <form
                  class=""
                >
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      data-testid="first-option-label"
                      for="first-option"
                    >
                      First Option
                    </label>
                    <input
                      class="form-control"
                      data-testid="first-option-input"
                      id="first-option"
                      placeholder="Option One"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      data-testid="second-option-label"
                      for="second-option"
                    >
                      Second Option
                    </label>
                    <input
                      class="form-control"
                      data-testid="second-option-input"
                      id="second-option"
                      placeholder="Option Two"
                      type="text"
                      value=""
                    />
                  </div>
                  <button
                    class="btn btn-light"
                    data-testid="poll-submit-btn"
                    type="submit"
                  >
                    Submit
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
